### PR TITLE
playerctl: add daemon service

### PR DESCRIPTION
**Pull request by aacebedo**
###### Description of changes

The playerctld service was not integrated in the playerctl package so I am adding it.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/desktops/playerctld.nix
Comment (SuperSandro2000) [inline nixos/modules/services/desktops/playerctld.nix:18]:
```suggestion
  options = {
    services.playerctld.enable = mkEnableOption (lib.mdDoc "Enable the playerctld daemon.");
  };
```

File: nixos/modules/services/desktops/playerctld.nix
Comment (SuperSandro2000) [inline nixos/modules/services/desktops/playerctld.nix:36]:
```suggestion
    environment.systemPackages = [ pkgs.playerctl ];

    systemd.user.services.playerctld = {
      description = "Playerctld daemon to track media player activity";
      wantedBy = [ "multi-user.target" ];

      serviceConfig = {
        Type = "oneshot";
        RemainAfterExit = true;
        ExecStart = "${pkgs.playerctl}/bin/playerctld daemon";
      };
    };
  };
```

Can the playerctld stay in the foreground? Then we don't need oneshot and remainAfterExit which are a bit ugly.

Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix:29]: yes it can. I modify this

Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix]: done

**Comment by aacebedo** [comment]
Thanks @SuperSandro2000 for the review. Applied the modifications

**Comment by aacebedo** [comment]
made a small change (added the possibility to select the package to install

File: nixos/modules/services/desktops/playerctld.nix
Comment (nagy) [inline nixos/modules/services/desktops/playerctld.nix:21]: `with pkgs;` is not needed in this context.

File: nixos/modules/services/desktops/playerctld.nix
Comment (nagy) [inline nixos/modules/services/desktops/playerctld.nix:14]: My reading of the `mkPackageOption` source code says that this should already be the default if it has the same name as above.

https://github.com/NixOS/nixpkgs/blob/ad7a8e6c27befbef28751586567027bedbcd72dc/lib/options.nix#L135

File: nixos/modules/services/desktops/playerctld.nix
Comment (nagy) [inline nixos/modules/services/desktops/playerctld.nix:30]: Would some hardening make sense here? Something like `DynamicUser=true` or `PrivateTmp=true` ? Because without a `User=` specification it would run as root, which should be avoided.

File: nixos/modules/services/desktops/playerctld.nix
Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix]: corrected

File: nixos/modules/services/desktops/playerctld.nix
Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix]: corrected

File: nixos/modules/services/desktops/playerctld.nix
Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix]: corrected

**Comment by aacebedo** [comment]
Pushed a fixup I will squash when everything is ok @nagy @SuperSandro2000

**Review by nagy — APPROVED** [review]

**Comment by aacebedo** [comment]
Squashed the fixup, shall be ready to merge

File: nixos/modules/services/desktops/playerctld.nix
Comment (SuperSandro2000) [inline nixos/modules/services/desktops/playerctld.nix:16]:
```suggestion
  options.services.playerctld = {
    enable = mkEnableOption (lib.mdDoc "the playerctld daemon");

    package = mkPackageOption pkgs "playerctl" { };
  };
```

File: nixos/modules/services/desktops/playerctld.nix
Comment (SuperSandro2000) [inline nixos/modules/services/desktops/playerctld.nix:34]:
```suggestion
    environment.systemPackages = [ cfg.package ];

    systemd.user.services.playerctld = {
      description = "Playerctld daemon to track media player activity";
      wantedBy = [ "multi-user.target" ];

      serviceConfig = {
        Type = "exec";
        ExecStart = "${cfg.package}/bin/playerctld";
      };
    };
  };

  meta = { maintainers = with lib.maintainers; [ aacebedo ]; };
```

**Comment by aacebedo** [comment]
added the suggestions.

**Review by SuperSandro2000 — COMMENTED** [review]
Please rebase, otherwise LGTM

**Comment by diniamo** [comment]
@aacebedo could you finish this PR? If not, then I'm reopening it.

**Comment by aacebedo** [comment]
Hi.
Forgot it sorry. I'll finish it tomorrow if you need it.

**Comment by diniamo** [comment]
Thanks, I'll add my own set of suggestions.

File: nixos/modules/services/desktops/playerctld.nix
Comment (diniamo) [inline nixos/modules/services/desktops/playerctld.nix:3]: This is a bad practice, generally you should avoid using with outside of package sets. Access the functions directly from lib, eg.: `lib.mkEnableOption`.

File: nixos/modules/services/desktops/playerctld.nix
Comment (diniamo) [inline nixos/modules/services/desktops/playerctld.nix:10]: `lib.mdDoc` has been removed from nixpkgs

File: nixos/modules/services/desktops/playerctld.nix
Comment (diniamo) [inline nixos/modules/services/desktops/playerctld.nix:29]: Use `meta.maintainers` please.

I would also argue that `with` should also be avoided here, since there is only one maintainer. But I'm not 100% sure since `with` is used in most places anyway, though I doubt that it's with good reason.

**Review by diniamo — CHANGES_REQUESTED** [review]

File: nixos/modules/services/desktops/playerctld.nix
Comment (diniamo) [inline nixos/modules/services/desktops/playerctld.nix:16]: The package does not need to be installed globally.

**Review by diniamo — CHANGES_REQUESTED** [review]

File: nixos/modules/services/desktops/playerctld.nix
Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix]: fixed

File: nixos/modules/services/desktops/playerctld.nix
Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix]: fixed

File: nixos/modules/services/desktops/playerctld.nix
Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix]: I let it in case other maintainers are added then

File: nixos/modules/services/desktops/playerctld.nix
Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix]: fixed

File: nixos/modules/services/desktops/playerctld.nix
Comment (diniamo) [inline nixos/modules/services/desktops/playerctld.nix:13]: The parentheses are reduant.

File: nixos/modules/services/desktops/playerctld.nix
Comment (diniamo) [reply nixos/modules/services/desktops/playerctld.nix]: Sorry, I take this back. It's done everywhere, and in this case, the daemon can also be controlled by subcommands.

File: nixos/modules/services/desktops/playerctld.nix
Comment (diniamo) [inline nixos/modules/services/desktops/playerctld.nix:21]: `multi-user.target` is for system services, it should be `default.target`

**Review by diniamo — CHANGES_REQUESTED** [review]

File: nixos/modules/services/desktops/playerctld.nix
Comment (nagy) [inline nixos/modules/services/desktops/playerctld.nix:21]: If you want to install this service as a user service, are you sure you want to use the `multi-user.target`? I only know this target from the system level instance. Does it transparently find, that this is what you want to depend upon? 

When I list the user targets, I see only these:

```
$ systemctl --user list-units --type=target
  UNIT           LOAD   ACTIVE SUB    DESCRIPTION
  basic.target   loaded active active Basic System
  default.target loaded active active Main User Target
  paths.target   loaded active active Paths
  sockets.target loaded active active Sockets
  timers.target  loaded active active Timers
```

Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix:21]: I think it is better to use default as you mentionned

Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix:21]: fixed

File: nixos/modules/services/desktops/playerctld.nix
Comment (aacebedo) [reply nixos/modules/services/desktops/playerctld.nix]: fixed

**Comment by diniamo** [comment]
@aacebedo in case you didn't see: you should install the package globally after all (see my reasoning in the original review message).

**Comment by aacebedo** [comment]
> @aacebedo in case you didn't see: you should install the package globally after all (see my reasoning in the original review thread).

yes I re-added it

**Comment by diniamo** [comment]
Ah, that's my bad, I can't read.

**Comment by diniamo** [comment]
LGTM then.

@SuperSandro2000 can you take a look and get this merged?

**Comment by diniamo** [comment]
@aacebedo looks like you have a trailing whitespace somewhere, that's why the CI is failing.

**Review by diniamo — APPROVED** [review]

File: nixos/modules/services/desktops/playerctld.nix
Comment (SuperSandro2000) [inline nixos/modules/services/desktops/playerctld.nix:23]: small question: this might be multi-user.target or graphical.target. Is that what you have intended?

File: nixos/modules/services/desktops/playerctld.nix
Comment (SuperSandro2000) [inline nixos/modules/services/desktops/playerctld.nix:20]:
```suggestion

```
We need to drop the extra spaces here to fix the editorcheck

File: nixos/modules/services/desktops/playerctld.nix
Comment (diniamo) [reply nixos/modules/services/desktops/playerctld.nix:23]: Notice that this is a user service, multi-user is for system services.

**Comment by diniamo** [comment]
@SuperSandro2000 could we get this merged?

**Review by h7x4 — APPROVED** [review]
I have tested functionality of the module, it works fine and the diff looks good

Would you mind adding a small release note?

**Review by nagy — APPROVED** [review]

**Comment by aacebedo** [comment]
@SuperSandro2000 can you merge please?

**Comment by aacebedo** [comment]
Can we merge this? Everything has been fixed

**Comment by h7x4** [comment]
@aacebedo I added a release note. Can you verify that it looks ok? Ping me once ofborg is happy for merge.

**Comment by diniamo** [comment]
I don't know if that's a big deal, but there is a typo in the commit message.

**Comment by aacebedo** [comment]
> I don't know if that's a big deal, but there is a typo in the commit message.

Yes maybe @h7x4 can fix it? A build seems blocked too.

**Comment by aacebedo** [comment]
@h7x4 you can merge

**Pull request merged**
